executor/aggregate_test.go

```diff
@@ -1142,3 +1145,127 @@ func (s *testSuiteAgg) TestIssue17216(c *C) {
 	tk.MustQuery("SELECT count(distinct col1) FROM t1").Check(testkit.Rows("48"))
 }
+
+func reconstructParallelGroupConcatResult(rows [][]interface{}) []string {
+	data := make([]string, 0, len(rows))
+	for _, row := range rows {
+		if str, ok := row[0].(string); ok {
+			tokens := strings.Split(str, ",")
+			sort.Slice(tokens, func(i, j int) bool {
+				return tokens[i] < tokens[j]
+			})
+			data = append(data, strings.Join(tokens, ","))
+		}
+	}
+
+	sort.Slice(data, func(i, j int) bool {
+		return data[i] < data[j]
+	})
+
+	return data
+}
+
+func (s *testSuiteAgg) TestParallelStreamAggGroupConcat(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("use test;")
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("CREATE TABLE t(a bigint, b bigint);")
+
+	for i := 0; i < 10000; i++ {
+		tk.MustExec("insert into t values(?, ?);", rand.Intn(100), rand.Intn(100))
+	}
+
+	sql := "select /*+ stream_agg() */ group_concat(a, b) from t group by b;"
+	concurrencies := []int{1, 2, 4, 8}
+	var expected []string
+	for _, con := range concurrencies {
+		tk.MustExec(fmt.Sprintf("set @@tidb_stream_agg_concurrency=%d", con))
+		if con == 1 {
+			expected = reconstructParallelGroupConcatResult(tk.MustQuery(sql).Rows())
+		} else {
+			er := tk.MustQuery("explain " + sql).Rows()
+			ok := false
+			for _, l := range er {
+				str := fmt.Sprintf("%v", l)
+				if strings.Contains(str, "Shuffle") {
+					ok = true
+					break
+				}
+			}
+			c.Assert(ok, Equals, true)
+			obtained := reconstructParallelGroupConcatResult(tk.MustQuery(sql).Rows())
+			c.Assert(len(obtained), Equals, len(expected))
+			for i := 0; i < len(obtained); i++ {
+				c.Assert(obtained[i], Equals, expected[i])
+			}
+		}
+	}
+}
+
+func (s *testSuiteAgg) TestIssue20658(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("use test;")
+
+	aggFuncs := []string{"count(a)", "sum(a)", "avg(a)", "max(a)", "min(a)", "bit_or(a)", "bit_xor(a)", "bit_and(a)"}
+	aggFuncs2 := []string{"var_pop(a)", "var_samp(a)", "stddev_pop(a)", "stddev_samp(a)", "approx_count_distinct(a)", "approx_percentile(a, 7)"}
```

Comment:
Why not combine them into one?

Reply:
for group_concat, it's hard to get reconstruct the result.